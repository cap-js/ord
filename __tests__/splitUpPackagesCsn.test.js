const cds = require("@sap/cds");
const path = require("path");
const { AUTHENTICATION_TYPE } = require("../lib/constants");
const ord = require("../lib/ord");
const bahCsn = require("./__mocks__/splitUpPackagesCsn.json");

describe("ORD Generation for Business Accelerator Hub", () => {
    beforeAll(() => {
        cds.root = path.join(__dirname, "bookshop");
        jest.spyOn(cds, "context", "get").mockReturnValue({
            authConfig: {
                types: [AUTHENTICATION_TYPE.Open],
            },
        });
        jest.spyOn(require("../lib/date"), "getRFC3339Date").mockReturnValue("2024-11-04T14:33:25+01:00");
    });

    beforeEach(() => {
        cds.env.ord = {
            namespace: "sap.test.cdsrc.sample",
            openResourceDiscovery: "1.10",
            description: "Business Accelerator Hub ORD Test",
            policyLevel: "sap:core:v1", //old value, check for compatibility
        };
    });

    afterAll(() => {
        jest.restoreAllMocks();
        jest.clearAllMocks();
    });

    describe("Tests for ORD document for Business Accelerator Hub", () => {
        test("Successfully create ORD Documents for Business Accelerator Hub", () => {
            const document = ord(bahCsn);
            expect(document).not.toBeUndefined();
            document.apiResources.forEach((api) => delete api.lastUpdate);
            document.eventResources.forEach((event) => delete event.lastUpdate);
            document.consumptionBundles.forEach((con) => delete con.lastUpdate);
            expect(document).toMatchSnapshot();
            expect(document.openResourceDiscovery).toBe("1.10");
            expect(document.policyLevels).toEqual(["sap:core:v1"]);
            expect(document.description).toBe("Business Accelerator Hub ORD Test");
            expect(document.apiResources).toHaveLength(7);
            expect(document.eventResources).toHaveLength(1);
        });
    });
});
