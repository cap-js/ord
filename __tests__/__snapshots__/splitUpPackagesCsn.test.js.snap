// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ORD Generation for Business Accelerator Hub Tests for ORD document for Business Accelerator Hub Successfully create ORD Documents for Business Accelerator Hub 1`] = `
{
  "$schema": "https://open-resource-discovery.github.io/specification/spec-v1/interfaces/Document.schema.json",
  "apiResources": [
    {
      "apiProtocol": "odata-v4",
      "description": "Description for SupplierService",
      "entryPoints": [
        "/odata/v4/supplier",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:SupplierService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:SupplierService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:SupplierService:v1/SupplierService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:SupplierService:v1/SupplierService.edmx",
        },
      ],
      "shortDescription": "Short description of SupplierService",
      "title": "SupplierService",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for MaterialService",
      "entryPoints": [
        "/odata/v4/material",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:MaterialService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:MaterialService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:MaterialService:v1/MaterialService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:MaterialService:v1/MaterialService.edmx",
        },
      ],
      "shortDescription": "Short description of MaterialService",
      "title": "MaterialService",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for PurchaseOrderService",
      "entryPoints": [
        "/odata/v4/purchase-order",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:PurchaseOrderService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:PurchaseOrderService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:PurchaseOrderService:v1/PurchaseOrderService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:PurchaseOrderService:v1/PurchaseOrderService.edmx",
        },
      ],
      "shortDescription": "Short description of PurchaseOrderService",
      "title": "PurchaseOrderService",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for InternalDataProductService",
      "entryPoints": [
        "/odata/v4/internal-data-product",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:InternalDataProductService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:InternalDataProductService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:InternalDataProductService:v1/InternalDataProductService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:InternalDataProductService:v1/InternalDataProductService.edmx",
        },
      ],
      "shortDescription": "Short description of InternalDataProductService",
      "title": "InternalDataProductService",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for CustomerService",
      "entryPoints": [
        "/odata/v4/customer",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:CustomerService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:CustomerService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:CustomerService:v1/CustomerService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:CustomerService:v1/CustomerService.edmx",
        },
      ],
      "shortDescription": "Short description of CustomerService",
      "title": "CustomerService",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for RestrictedDataProductService",
      "entryPoints": [
        "/odata/v4/restricted-data-product",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:RestrictedDataProductService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:RestrictedDataProductService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:RestrictedDataProductService:v1/RestrictedDataProductService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:RestrictedDataProductService:v1/RestrictedDataProductService.edmx",
        },
      ],
      "shortDescription": "Short description of RestrictedDataProductService",
      "title": "RestrictedDataProductService",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "Description for BusinessAcceleratorHub",
      "entryPoints": [
        "/odata/v4/business-accelerator-hub",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:BusinessAcceleratorHub:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:BusinessAcceleratorHub",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:BusinessAcceleratorHub:v1/BusinessAcceleratorHub.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/ord/v1/sap.test.cdsrc.sample:apiResource:BusinessAcceleratorHub:v1/BusinessAcceleratorHub.edmx",
        },
      ],
      "shortDescription": "Short description of BusinessAcceleratorHub",
      "title": "BusinessAcceleratorHub",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "consumptionBundles": [
    {
      "description": "This Consumption Bundle contains all resources of the reference app which are unprotected and do not require authentication",
      "ordId": "capirebookshopordsample:consumptionBundle:noAuth:v1",
      "shortDescription": "If we have another protected API then it will be another object",
      "title": "Unprotected resources",
      "version": "1.0.0",
    },
  ],
  "description": "Business Accelerator Hub ORD Test",
  "eventResources": [
    {
      "description": "CAP Event resource describing events / messages.",
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:eventResource:BusinessAcceleratorHub:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:BusinessAcceleratorHub",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-event:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/ord/v1/sap.test.cdsrc.sample:eventResource:BusinessAcceleratorHub:v1/BusinessAcceleratorHub.asyncapi2.json",
        },
      ],
      "shortDescription": "BusinessAcceleratorHub event resource",
      "title": "ODM capirebookshopordsample Events",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "groups": [
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:SupplierService",
      "groupTypeId": "sap.cds:service",
      "title": "Supplier Service",
    },
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:MaterialService",
      "groupTypeId": "sap.cds:service",
      "title": "Material Service",
    },
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:PurchaseOrderService",
      "groupTypeId": "sap.cds:service",
      "title": "PurchaseOrder Service",
    },
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:InternalDataProductService",
      "groupTypeId": "sap.cds:service",
      "title": "InternalDataProduct Service",
    },
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:CustomerService",
      "groupTypeId": "sap.cds:service",
      "title": "Customer Service",
    },
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:RestrictedDataProductService",
      "groupTypeId": "sap.cds:service",
      "title": "RestrictedDataProduct Service",
    },
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:BusinessAcceleratorHub",
      "groupTypeId": "sap.cds:service",
      "title": "BusinessAcceleratorHub Service",
    },
  ],
  "openResourceDiscovery": "1.10",
  "packages": [
    {
      "description": "This package contains public APIs for capire bookshop ord sample.",
      "ordId": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "shortDescription": "Package containing public APIs",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "1.0.0",
    },
    {
      "description": "This package contains public Events for capire bookshop ord sample.",
      "ordId": "sap.test.cdsrc.sample:package:capirebookshopordsample-event:v1",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "shortDescription": "Package containing public Events",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "1.0.0",
    },
  ],
  "policyLevels": [
    "sap:core:v1",
  ],
  "products": [
    {
      "ordId": "customer:product:capire.bookshop.ord.sample:",
      "shortDescription": "Short description of capire bookshop ord sample",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
    },
  ],
}
`;
